src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2.cs

```diff
@@ -530,5 +531,114 @@ public MemoryPoolIterator2 CopyTo(byte[] array, int offset, int count, out int a
                 }
             }
         }
+
+        public int CopyFrom(ArraySegment<byte> buffer)
+        {
+            Debug.Assert(_block != null);
+            Debug.Assert(_block.Pool != null);
+            Debug.Assert(_block.Next == null);
+            Debug.Assert(_block.End == _index);
+
+            var pool = _block.Pool;
+            var block = _block;
+            var blockIndex = _index;
+
+            var bufferIndex = buffer.Offset;
+            var remaining = buffer.Count;
+
+            while (remaining > 0)
+            {
+                var bytesLeftInBlock = block.Data.Offset + block.Data.Count - blockIndex;
+
+                if (bytesLeftInBlock == 0)
+                {
+                    var nextBlock = pool.Lease();
+                    block.Next = nextBlock;
+                    block = nextBlock;
+
+                    blockIndex = block.Data.Offset;
+                    bytesLeftInBlock = block.Data.Count;
+                }
+
+                var bytesToCopy = Math.Min(remaining, bytesLeftInBlock);
+
+                Buffer.BlockCopy(buffer.Array, bufferIndex, block.Array, blockIndex, bytesToCopy);
+
+                blockIndex += bytesToCopy;
+                bufferIndex += bytesToCopy;
+                remaining -= bytesToCopy;
+                block.End = blockIndex;
+            }
+
+            _block = block;
+            _index = blockIndex;
+            return buffer.Count;
+        }
+
+        public int CopyFrom(byte[] data)
+        {
+            return CopyFrom(new ArraySegment<byte>(data));
+        }
+
+        public int CopyFrom(byte[] data, int offset, int count)
+        {
+            return CopyFrom(new ArraySegment<byte>(data, offset, count));
+        }
+
+        public unsafe int CopyFromAscii(string data)
+        {
+            Debug.Assert(_block.Next == null);
+            Debug.Assert(_block.End == _index);
+
+            var block = _block;
+
+            var inputLength = data.Length;
+            var inputLengthMinusSpan = inputLength - 3;
+
+            fixed (char* pData = data)
+            {
+                var input = pData;
+                var inputEnd = pData + data.Length;
+                var blockRemaining = block.Data.Offset + block.Data.Count - block.End;
+                var blockRemainingMinusSpan = blockRemaining - 3;
+
+                while (input < inputEnd)
+                {
+                    if (blockRemaining == 0)
+                    {
+                        block.Next = block.Pool.Lease();
+                        block = block.Next;
+                        blockRemaining = block.Data.Count;
+                        blockRemainingMinusSpan = blockRemaining - 3;
+                    }
+
+                    fixed (byte* pOutput = block.Data.Array)
```

Comment:
Have a PR with results for that https://github.com/aspnet/KestrelHttpServer/pull/399#issuecomment-157728877 main issue is casting char to byte; or in the referenced PR byte to char; else could use block copy.